torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -124,11 +142,53 @@ def _pre_forward(
     def _post_forward(self, module: nn.Module, input: Any, output: Any) -> Any:
         if self._fsdp_param_group:
             output = self._fsdp_param_group.post_forward(module, input, output)
+        output = self._register_pre_backward_hook(output)
         self._training_state = TrainingState.IDLE
-        if self._is_root and (all_gather_state := self._all_gather_state.pop()):
-            self._all_gather_copy_in_stream.wait_event(all_gather_state.event)
-            self._all_gather_stream.wait_event(all_gather_state.event)
-            del all_gather_state  # free
+        if self._state_ctx.iter_forward_root is self:
+            if all_gather_state := self._comm_ctx.all_gather_state:
```

Comment:
add some doc about why here the copy in stream need to wait for the allgather event?

Reply:
I added a note in `_fsdp_param_group.py` to explain how we overlap copy-in and all-gather, and I added a reference to that note here.